### ISSUE TITLE
More generic if_else possible?

### ISSUE THREAD

**Issue by omersahintas** [casadi/casadi#2766]
Dear Joel & Joris,

Thanks for developing and maintaining casadi!

I have a special case, in which I have to process some data segments in my minimization problem. Which data segment I have to pick depends on the values of the parameters (no worries, segments have continuous derivatives). Therefore, I have to iterate all the data segments to find the appropriate one during optimization.

Now the problem arises: the index of the data segment in the vector is `int` (or `size_t`). However, `if_else` templates the output to the type of the condition. Please refer to these lines:

https://github.com/casadi/casadi/blob/12fa60f676716ae09aa2abe34c1c9b9cf426e68a/casadi/core/matrix_impl.hpp#L2379
https://github.com/casadi/casadi/blob/12fa60f676716ae09aa2abe34c1c9b9cf426e68a/casadi/core/matrix_decl.hpp#L327

I was just considering if it is possible to implement an `if_else` block, that doesn't do any type conversion.

For my specific application, I would iterate over the data segment vector and check the condition for a symbolic type, but store or return the appropriate index.

If you have any other suggestions on how to deal with this case (maybe using a modified callback? -- However my objective function consists of multiple terms and I would very much like to use full hessian), I would be glad if you would mention these.

Thanks!

Best,
Sahin

**Comment by jaeandersson** [comment]
I'm not following. What kind of type conversion are you talking about? Do you mean the `Matrix<Scalar>` return type for `Matrix<Scalar>::if_else`? The condition is a symbolic expression so the output has to be a symbolic expression.

**Comment by omersahintas** [comment]
Yes, I meant the `Matrix<Scalar>` return type for `Matrix<Scalar>::if_else`. My intention was to access a variable inside the  `Matrix<Scalar>::if_else` from an "outer" scope.  But that does not make sense while using symbolic expressions.

I will close this issue. Thanks for your reply!
